pisa/utils/config_parser.py

```diff
@@ -614,6 +617,26 @@ def parse_pipeline_config(config):
     # Get and parse the order of the stages (and which services implement them)
     order = [split(x, STAGE_SEP) for x in split(config.get(section, 'order'))]
 
+
+    #
+    # Browse through the list of (service, stage) Tuple to skip, if any
+    #
+    if skip_stages is not None:
+        # Check the types of skip_Stages. must be a list of Tuples
+        assert isinstance(skip_stages,list) and isinstance(skip_stages[0],tuple), 'ERROR: skip_stages must be a list of Tuples'
```

Comment:
Assertion errors are less useful than raising a TypeError in the case that types are incorrect for other people's code and for unit testing. But input shouldn't need to be a list, just the more generic abstract base class (abc) `Sequence`. And why do the inner sequences have to be tuples? Why should a user expect a list of two items specifying `[stage, service]` to be illegal?

Also, does a user specifying a single (stage, service) have to pass that as [(stage, service)]? Why not simplify the user's life, and allow `(stage, service)` or `[(stage, service)]` or `[(stage, service), (stage, service), ...]`? Code below should implement this in a fairly concise manner


```python
# at the top of the file in the "python builtin imports" section
from collections.abc import Sequence


# ... inside your function, preferably at the top to

if not (skip_services is None or isinstance(skip_services, Sequence)):
    raise TypeError(
        "`skip_services` expected to be None or Sequence, got {}".format(type(skip_services))
    )

if skip_services is None:
    skip_services = []
else:
    if isinstance(skip_services[0], str):
        skip_services = [skip_services]

    normalized_skip_services = []

    for skip_service in skip_services:
        if not isinstance(skip_service, Sequence):
            raise TypeError("expected Sequence, got {}".format(type(skip_service))

        if len(skip_service) != 2:
            raise TypeError(
                '(stage, service) sequences be length 2; got "{}", of length {}'.format(
                    skip_service, len(skip_service)
                )
            )

        for item, name in zip(skip_service, ["Stage", "Service"]:
            if not isinstance(item, str):
                raise TypeError(
                    "{} must be specified as a string; got {}".format(name, type(item))
                )

        normalized_skip_services.append((skip_service[0], skip_service[1]))

    skip_services = normalized_skip_services
```

Reply:
the restriction to tuples was made so that the pairs could be easily compared within the two lists (I initially used sets type to find the difference between two sets of tuples, but it didn't preserve the order). You're right that with the way things are implemented now, the type of the skip_service can be a bit more flexible.

Adding various type options like this sounds good.